### PR TITLE
Add support for Energy Blade

### PR DESCRIPTION
**Pull request by PJacek**
This was a bit complicated because Energy Blade actually replaces weapons (the bases exist in data, not doing some magic like "treat x stat as y"). Some skills require the use of certain weapon types, so items need to be processed before skills. Energy Blade needs to know if it's active when replacing weapons, so skills need to be processed before items. This creates circular dependency, which I think can only be solved with a config option (processed at the very start).
- Added the special bases - they have modified names because both are called "Energy Blade" in GGPK and we export all bases into the same `itemBases` table. They are not available for crafting
- Used weapons are completely replaced with Energy Blade(s)
  - Requirements of the original weapon are used - the swap happens later and energy blades have no requirements
  - Any mods present on original weapon are discarded
  - Weapon type restrictions function correctly - Close Combat can be used, mods requiring Sword attacks apply, etc.
- Crit chance from quality applies to base crit as if it was a local mod
- More damage bonus when using a two handed weapon applies to the base damage
- Battlemage works correctly
- Weapon restrictions can now be applied to non-attack skills. This disables the skill (and buff) if no usable weapon is equipped
### Sample build
```
https://pastebin.com/w5x1x15n
```
### Screenshots
Energy Blade buff:
image
Concentrated Path supported by Close Combat usable with a staff equipped:
image
Base crit increased by quality and sword mods applying:
image
Skill disabled when using a bow or unarmed:
image
Requirements from the original weapon apply but the mods don't:
image
Battlemage:
image

### CODE REVIEW

**Comment by robinerd** [comment]
Hi! Thanks for adding this support! I'm currently running a build with energy blade and wanted to try it out in PoB. Found this commit but have never looked at the source code of PoB before. Also want to point out some issues/findings I came across.

I managed to setup the development installation, took the dev branch and merged in the branch from your repo too. I also exported the game data with the .dat tool thingy.

The first issue I had was that I wasn't getting the energy blade weapon, but then I found it's since when I import my character it doesn't import any weapon on my mainhand, and so, I won't be able to get the energy blade buff since that gem requires a weapon.
If I manually edit my build in PoB to equip any 1H weapon, I get the buff and all works.

My second issue was that it also doesn't import any of the gems in my main hand. I suspect that something needs to be changed in the import code to correctly import my character if energy blade is turned on in-game.

NOTE: it could also just be something I haven't setup properly in the dev version, so please ignore me if that's the case.

Anyway as a workaround I can just manually add my last gem and add any 1H weapon, and I will be able to test things with my build. Cheers and thanks again!

File: src/Modules/CalcSetup.lua [hunk 1]
```diff
@@ -621,6 +621,36 @@ function calcs.initEnv(build, mode, override, specEnv)
 							env.itemModDB:ScaleAddMod(mod, scale)
 						end
 					end
+				elseif (slotName == "Weapon 1" or slotName == "Weapon 2") and modDB:Flag(nil, "Condition:EnergyBladeActive") then
```
Comment (Wires77) [inline src/Modules/CalcSetup.lua:624]: Would it not be easier to just use item:ParseRaw(```
test
Energy Blade One Handed
Quality: 20
Sockets: G-G-G
Implicits: 1```)

and add the raw implicit text from there?

**Review by Wires77 — CHANGES_REQUESTED** [review]
Looks like the import bug is because the sword imported looks like this:
```
Rarity: Magic
Superior Energy Blade
--------
Item Level: 60
--------
One Handed Sword
Elemental Damage: 56-1120
Critical Strike Chance: 8.54%
Attacks per Second: 1.70
Weapon Range: 11
--------
Requirements:
Level: 70
Str: 155
Dex: 116
--------
Sockets: R-R-G
--------
Adds 56 to 1120 Lightning Damage (implicit)
--------
22% increased Critical Strike Chance
```

That base type doesn't exist (due to the renaming you had to do).  If we can fix that, or at least import some dummy weapon when we see the "Energy Blade" base, this should be good

File: src/Classes/ImportTab.lua [hunk 2]
```diff
@@ -628,6 +628,17 @@ function ImportTabClass:ImportItem(itemData, slotName)
 		end
 	else
 		item.name = itemLib.sanitiseItemText(itemData.typeLine)
```
Comment (Wires77) [inline src/Classes/ImportTab.lua:630]: Looks like this ends up being "Superior Energy Blade" if there's quality on the gem.  We could use `itemData.baseType` instead, but I'm not sure what else that would affect, as far as the typeLine goes

**Review by Wires77 — APPROVED** [review]

**Pull request merged**